### PR TITLE
w3c tracestate: Tests P

### PR DESCRIPTION
**Pull request by mabdinur**
## Motivation

- Tests new datadog member in tracestate

## Changes

- adds tests for new Datadog tracestate field and dd.parent id tag

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
    * [ ] To R&P team: locally build and push the image to hub.docker.com 
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
    * [ ] Once merged, add (or remove) it in system-test-dasboard nightly

### CODE REVIEW

File: tests/parametric/test_headers_tracecontext.py [hunk 1]
```diff
@@ -713,6 +713,84 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
         assert traceparent4.trace_id == "12345678901234567890123456789012"
         assert "foo=1" in str(tracestate4) or "foo=2" in str(tracestate4)
 
+    def test_tracestate_w3c_lp_id_extract(self, test_agent, test_library):
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:716]:
```suggestion
    @missing_feature(context.library == "java", reason="Not implemented")
    @missing_feature(context.library="python", reason="Not implemented")
    @missing_feature(context.library="dotnet", reason="Not implemented")
    @missing_feature(context.library="php", reason="Not implemented")
    @missing_feature(context.library="nodejs", reason="Not implemented")
    @missing_feature(context.library="dotnet", reason="Not implemented")
    @missing_feature(context.library="php", reason="Not implemented")
    def test_tracestate_w3c_lp_id_extract(self, test_agent, test_library):
```

File: tests/parametric/test_headers_tracecontext.py [hunk 2]
```diff
@@ -713,6 +713,84 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
...
+
+        assert case3["name"] == "datadog_headers_used_in_propagation"
+        assert case3["trace_id"] == 5
+        assert case3["parent_id"] == 11
+        assert "_dd.lp.id" not in case3["meta"]
+
+        assert case4["name"] == "lp_id_not_propagated"
+        assert case4["meta"]["_dd.lp.id"] == "0000000000000000"
+
+    def test_tracestate_w3c_lp_id_inject(self, test_agent, test_library):
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:778]:
```suggestion
    @missing_feature(context.library == "java", reason="Not implemented")
    @missing_feature(context.library == "python", reason="Not implemented")
    @missing_feature(context.library == "dotnet", reason="Not implemented")
    @missing_feature(context.library == "php", reason="Not implemented")
    @missing_feature(context.library == "nodejs", reason="Not implemented")
    @missing_feature(context.library == "dotnet", reason="Not implemented")
    @missing_feature(context.library == "php", reason="Not implemented")
    def test_tracestate_w3c_lp_id_inject(self, test_agent, test_library):
```

File: tests/parametric/test_headers_tracecontext.py [hunk 3]
```diff
@@ -713,6 +713,84 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
         assert traceparent4.trace_id == "12345678901234567890123456789012"
         assert "foo=1" in str(tracestate4) or "foo=2" in str(tracestate4)
 
+    def test_tracestate_w3c_lp_id_extract(self, test_agent, test_library):
+        """
+        Ensure the last parent id tag is set according to the W3C spec
+        """
+        with test_library:
+            with test_library.start_span(
+                name="lp_id_set",
```
Comment (bouwkast) [inline tests/parametric/test_headers_tracecontext.py:722]: nitpicky, but maybe we'd want to change the names from `lp_id_*` to just `p_*`?

**Review by bouwkast — APPROVED** [review]
Looks good to me

Only nitpicky thing is maybe changing the names of the spans, but not a blocker

File: tests/parametric/test_headers_tracecontext.py [hunk 4]
```diff
@@ -713,6 +713,96 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
...
+
+        assert case4["name"] == "p_not_propagated"
+        assert case4["meta"]["_dd.parent_id"] == "0000000000000000"
+
+    @missing_feature(context.library == "python", reason="Not implemented")
+    @missing_feature(context.library == "dotnet", reason="Not implemented")
+    @missing_feature(context.library == "php", reason="Not implemented")
+    @missing_feature(context.library == "nodejs", reason="Not implemented")
+    @missing_feature(context.library == "dotnet", reason="Not implemented")
+    @missing_feature(context.library == "php", reason="Not implemented")
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:789]:
```suggestion
    @missing_feature(context.library == "nodejs", reason="Not implemented")
    @missing_feature(context.library == "java", reason="Not implemented")
    @missing_feature(context.library == "cpp", reason="Not implemented")
    @missing_feature(context.library == "ruby", reason="Not implemented")
```

File: tests/parametric/test_headers_tracecontext.py [hunk 5]
```diff
@@ -713,6 +713,100 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
...
+        traces = test_agent.wait_for_num_traces(4)
+
+        assert len(traces) == 4
+        case1, case2, case3, case4 = traces[0][0], traces[1][0], traces[2][0], traces[3][0]
+
+        assert case1["name"] == "p_set"
+        assert case1["meta"]["_dd.parent_id"] == "0123456789abcdef"
+
+        assert case2["name"] == "p_invalid"
+        assert case2["meta"]["_dd.parent_id"] == "XX!X", f"{case2}"
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:776]:
```suggestion
        assert case2["meta"]["_dd.parent_id"] == "XX!X"
```

File: tests/parametric/test_headers_tracecontext.py [hunk 6]
```diff
@@ -713,6 +713,100 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
...
 
+    @missing_feature(context.library == "python", reason="Not implemented")
+    @missing_feature(context.library == "dotnet", reason="Not implemented")
+    @missing_feature(context.library == "php", reason="Not implemented")
+    @missing_feature(context.library == "nodejs", reason="Not implemented")
+    @missing_feature(context.library == "java", reason="Not implemented")
+    @missing_feature(context.library == "cpp", reason="Not implemented")
+    @missing_feature(context.library == "ruby", reason="Not implemented")
+    @missing_feature(context.library == "golang", reason="Not implemented")
+    def test_tracestate_w3c_p_extract(self, test_agent, test_library):
```
Comment (erikayasuda) [inline tests/parametric/test_headers_tracecontext.py:724]: nit: Are we not parametrizing this test because of case 3?

**Review by erikayasuda — APPROVED** [review]
One non-blocker question, otherwise LGTM

File: tests/parametric/test_headers_tracecontext.py (hunk 6 above)
Comment (mabdinur) [reply tests/parametric/test_headers_tracecontext.py:724]: Currently this feature is not merged and released in any tracers so we have to skip this test. The plan is to enable these tests by the EOQ

Comment (mabdinur) [reply tests/parametric/test_headers_tracecontext.py:724]: Case 3 doesn't have anything to do with this

**Comment by mabdinur** [comment]
The failing tracestate parametric test is unrelated.

File: tests/parametric/test_headers_tracecontext.py [hunk 7]
```diff
@@ -713,6 +713,100 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
         assert traceparent4.trace_id == "12345678901234567890123456789012"
         assert "foo=1" in str(tracestate4) or "foo=2" in str(tracestate4)
 
+    @missing_feature(context.library == "python", reason="Not implemented")
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:716]:
```suggestion
    @missing_feature(context.library < "python@2.7.0", reason="Not implemented")
```

File: tests/parametric/test_headers_tracecontext.py [hunk 8]
```diff
@@ -713,6 +713,100 @@ def test_tracestate_duplicated_keys(self, test_agent, test_library):
...
+
+        assert case3["name"] == "datadog_headers_used_in_propagation"
+        assert case3["trace_id"] == 5
+        assert case3["parent_id"] == 11
+        assert "_dd.parent_id" not in case3["meta"]
+
+        assert case4["name"] == "p_not_propagated"
+        assert case4["meta"]["_dd.parent_id"] == "0000000000000000"
+
+    @missing_feature(context.library == "python", reason="Not implemented")
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:786]:
```suggestion
    @missing_feature(context.library < "python@2.7.0", reason="Not implemented")
```

**Review by robertomonteromiguel — APPROVED** [review]

**Pull request merged**